src/Functions/FunctionsConversion.cpp

```diff
@@ -3736,20 +3737,41 @@ class FunctionCast final : public IFunctionBase
             element_wrappers.reserve(to_names.size());
             to_reverse_index.reserve(from_names.size());
 
+            size_t num_from_fields = 0;
             for (size_t i = 0; i < to_names.size(); ++i)
             {
                 auto it = from_positions.find(to_names[i]);
                 if (it != from_positions.end())
                 {
                     element_wrappers.emplace_back(prepareUnpackDictionaries(from_element_types[it->second], to_element_types[i]));
                     to_reverse_index.emplace_back(it->second);
+                    ++num_from_fields;
                 }
                 else
                 {
                     element_wrappers.emplace_back();
                     to_reverse_index.emplace_back();
                 }
             }
+
+            bool strict_named_tuple_conversion = false;
+
+            if (DB::CurrentThread::isInitialized())
+            {
+                const DB::ContextPtr query_context = DB::CurrentThread::get().getQueryContext();
+
+                // Avoid strict checks for background operations like part
+                // merging to achieve better backward compatibility.
+                if (query_context && !query_context->isBackgroundOperationContext())
+                    strict_named_tuple_conversion = query_context->getSettingsRef()[Setting::strict_named_tuple_conversion];
+            }
+
+            if (strict_named_tuple_conversion && num_from_fields < from_names.size())
```

Comment:
Let's move this to the for loop instead so we can fail early with the proper column name:

```
            for (size_t i = 0; i < to_names.size(); ++i)
            {
                auto it = from_positions.find(to_names[i]);
                if (it != from_positions.end())
                {
                    element_wrappers.emplace_back(prepareUnpackDictionaries(from_element_types[it->second], to_element_types[i]));
                    to_reverse_index.emplace_back(it->second);
                }
                else
                {
                    if (strict_named_tuple_conversion)
                        throw Exception(
                            ErrorCodes::CANNOT_CONVERT_TYPE,
                            "Missing field '{}' when casting {} to {}",
                            to_names[i],
                            from_type->getName(),
                            to_type->getName());
                    element_wrappers.emplace_back();
                    to_reverse_index.emplace_back();
                }
            }
```

This way we throw an error like `Code: 70. DB::Exception: Missing field 'a' when casting Tuple(a2 UInt8, b2 UInt8) to Tuple(a Int32, b Int32)`

Reply:
> Let's move this to the for loop instead so we can fail early with the proper column name:

But this is not desired. It's acceptable to convert `Tuple(a int)` to `Tuple(a int, b int)`.

Reply:
With `strict_named_tuple_conversion` on I would expect `Select (0)::Tuple(a int)::Tuple(a int, b int)` to throw, because that's the whole point of having a strict validation.

For example, `Select (0)::Tuple(int)::Tuple(int, int)` currently throws because due to `CAST AS Tuple can only be performed between tuple types with the same number of elements or from String`.

Reply:
> because that's the whole point of having a strict validation.

I'm afraid not. The motivation is to throw exception when converting `Tuple(a int, b int)` into `Tuple(a int)`, which leads to data loss. It's totally fine to add more fields into existing tuple type.

Reply:
> It's totally fine to add more fields into existing tuple type.

Why? We don't consider it fine when the tuple elements are unnamed. In fact it leads to this situation:
```
Select (42)::Tuple(int)::Tuple(int, int); -- KO
Select (42)::Tuple(int)::Tuple(a int, b int); -- KO
Select (42)::Tuple(int)::Tuple(a int)::Tuple(a int, b int); -- OK
Select (42)::Tuple(int)::Tuple(a int)::Tuple(a int, b int); -- KO
Select (42)::Tuple(int)::Tuple(a int)::Tuple(a int, b int)::Tuple(int, int); -- OK
```

I'd expect for both named and unnamed tuples to work the same way (either both can add new members or both can't), and since doing odd things during casts has already proven confusing and dangerous I think the default should be sane and throw an error. If we need a custom cast function let's add it (we already have dozens of them, like  `toInt32OrDefault`) but it's better to keep cast as unsurprising as possible.

Reply:
> We don't consider it fine when the tuple elements are unnamed

That's unfortunate tbh. But I get your idea.

However, it's likely to break more stuff if we don't allow adding new fields into named tuple by default. I'll try.

Reply:
> 02887_mutations_subcolumns

New failures when we make it stricter. After a second thought, I think we should only consider named tuple conversions and leave out unnamed tuples, as the original motivation is to add a safe guard for the feature `enable_named_columns_in_function_tuple`. Converting `Tuple(a int)` to `Tuple(a int, b int)` is quite natural when modeling JSON data in ClickHouse (without the newly introduced JSON type), and it doesn't lead to data loss.